File: Assets/Resources/RGSS/tilemap.rb
require 'type_check_util'

class Tilemap
  include TypeCheckUtil

  attr_reader :__handler__

  def initialize(viewport = nil)
    @__handler__ = Unity::Tilemap.new(viewport)
  end

  [:dispose, :disposed?, :update].each do |method|
    define_method(method) { @__handler__.send(method) }
  end

  def bitmaps
    @__handler__.bitmaps.map { |bitmap| Unity::Bitmap.new(bitmap) }
  end

  def map_data
    @__handler__.map_data
  end

  def map_data=(map_data)
    check_arguments([map_data], [Table])
    @__handler__.map_data = map_data.__handler__
  end

  def flash_data
    @__handler__.flash_data
  end

  def flash_data=(flash_data)
    check_arguments([flash_data], [Table])
    @__handler__.flash_data = flash_data.__handler__
  end

  def flags
     @__handler__.flags
  end

  def flags=(flags)
    check_arguments([flags], [Table])
    @__handler__.flags = flags.__handler__
  end

  def viewport
    @__handler__.viewport
  end

  def viewport=(viewport)
    check_arguments([viewport], [Viewport])
    @__handler__.viewport = viewport.__handler__
  end

  def eql?(other)
    if self == other
      true
    end
    self.__handler__ == other.__handler__
  end

  def hash
    @__handler__.hash
  end

  [:ox, :oy, :visible].each do |prop|
    define_method(prop) { @__handler__.send(prop) }
    define_method("#{prop}=") { |value| @__handler__.send("#{prop}=", value) }
  end
end

File: Assets/Scripts/RubyClasses/Tilemap.cs
namespace RGSSUnity.RubyClasses
{
    using MRuby.Library.Language;
    using MRuby.Library.Mapper;

    [RbClass("Tilemap", "Object", "Unity")]
    public static class Tilemap
    {
        [RbClassMethod("new_with_viewport")]
        public static RbValue NewWithViewport(RbState state, RbValue self, RbValue viewport)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("dispose")]
        public static RbValue Dispose(RbState state, RbValue self)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("disposed?")]
        public static RbValue Disposed(RbState state, RbValue self)
        {
            return state.RbFalse;
        }
        
        [RbInstanceMethod("update")]
        public static RbValue Update(RbState state, RbValue self)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("viewport")]
        public static RbValue Viewport(RbState state, RbValue self)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("map_data")]
        public static RbValue MapData(RbState state, RbValue self)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("map_data=")]
        public static RbValue MapDataSet(RbState state, RbValue self, RbValue value)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("flash_data")]
        public static RbValue FlashData(RbState state, RbValue self)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("flash_data=")]
        public static RbValue FlashDataSet(RbState state, RbValue self, RbValue value)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("bitmaps")]
        public static RbValue Bitmaps(RbState state, RbValue self)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("bitmaps=")]
        public static RbValue BitmapsSet(RbState state, RbValue self)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("flags")]
        public static RbValue Flags(RbState state, RbValue self)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("flags=")]
        public static RbValue FlagsSet(RbState state, RbValue self, RbValue value)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("visible")]
        public static RbValue Visible(RbState state, RbValue self)
        {
            return state.RbFalse;
        }
        
        [RbInstanceMethod("visible=")]
        public static RbValue VisibleSet(RbState state, RbValue self, RbValue value)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("ox")]
        public static RbValue Ox(RbState state, RbValue self)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("ox=")]
        public static RbValue OxSet(RbState state, RbValue self, RbValue value)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("oy")]
        public static RbValue Oy(RbState state, RbValue self, RbValue value)
        {
            return state.RbNil;
        }
        
        [RbInstanceMethod("oy=")]
        public static RbValue OySet(RbState state, RbValue self, RbValue value)
        {
            return state.RbNil;
        }
    }
}